packages/client/src/components/app/forms/RelationshipField.svelte

```diff
@@ -50,54 +36,51 @@
       limit: 100,
     },
   })
-  $: hasFilter = !!filter?.filter(f => !!f.field)?.length
-  $: fetch.update({ filter })
-  $: {
-    options = searchResults ? searchResults : $fetch.rows
-    const nonMatchingOptions = selectedOptions.filter(
-      option => !options.map(opt => opt._id).includes(option._id)
-    )
-    // Append initially selected options if there is no filter
-    // and hasn't already been appended
-    if (!hasFilter) {
-      options = [...options, ...nonMatchingOptions]
-    }
-  }
+
   $: tableDefinition = $fetch.definition
-  $: primaryDisplay = tableDefinition?.primaryDisplay || "_id"
-  $: singleValue = flatten(fieldState?.value)?.[0]
-  $: multiValue = flatten(fieldState?.value) ?? []
+  $: selectedValue = multiselect
+    ? flatten(fieldState?.value) ?? []
+    : flatten(fieldState?.value)?.[0]
   $: component = multiselect ? CoreMultiselect : CoreSelect
   $: expandedDefaultValue = expand(defaultValue)
-  $: debouncedSearch(searchString)
+  $: primaryDisplay = tableDefinition?.primaryDisplay
+
+  let optionsObj = {}
+  let initialValuesProcessed
   $: {
-    if (searching) {
-      isOpen = true
+    if (!initialValuesProcessed && primaryDisplay) {
+      initialValuesProcessed = true
+      optionsObj = {
+        ...optionsObj,
+        ...fieldState?.value?.reduce((accumulator, value) => {
+          accumulator[value._id] = {
+            _id: value._id,
+            [primaryDisplay]: value.primaryDisplay,
+          }
+          return accumulator
+        }, {}),
+      }
     }
   }
+  $: optionsObj = {
+    ...optionsObj,
+    ...($fetch.rows || [])?.reduce((accumulator, row) => {
+      accumulator[row._id] = row
+      return accumulator
+    }, {}),
+  }
 
-  // Fetch the initially selected values
-  // as they may not be within the first 100 records
-  $: {
-    if (
-      primaryDisplay !== "_id" &&
-      fieldState?.value?.length &&
-      !selectedOptions?.length
-    ) {
-      API.searchTable({
-        paginate: false,
-        tableId: linkedTableId,
-        limit: 100,
-        query: {
-          oneOf: {
-            [`1:${primaryDisplay}`]: fieldState?.value?.map(
-              value => value.primaryDisplay
-            ),
-          },
-        },
-      }).then(response => {
-        const value = multiselect ? multiValue : singleValue
-        selectedOptions = response.rows.filter(row => value.includes(row._id))
+  $: fetchRows(searchTerm, primaryDisplay)
```

Comment:
Thanks for breaking some of these giant reactive statements out into functions - it's much clearer to read and also gives us much more explicit control over when svelte reactivity will invoke them.